File: client_side.py
import socket
import time
import argparse

# Argument parsing to support reverse mode and port selection
parser = argparse.ArgumentParser(description="Simple iperf3-like client with reverse mode and logging")
parser.add_argument('-s', '--server', type=str, required=True, help="Server IP address")
parser.add_argument('-p', '--port', type=int, default=5201, help="Server port (default 5201)")
parser.add_argument('-t', '--time', type=int, default=10, help="Duration of data transfer in seconds")
parser.add_argument('-R', '--reverse', action='store_true', help="Enable reverse mode (server sends data to client)")

args = parser.parse_args()

SERVER_HOST = args.server
SERVER_PORT = args.port
BUFFER_SIZE = 128 * 1024  # 128 KB buffer size
DATA = b'X' * BUFFER_SIZE  # Data to be sent in normal mode
DOWNLOAD_DURATION = args.time
LOG_INTERVAL = 1  # Log every 1 second

# Create a TCP socket
client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

# Connect to the server
client_socket.connect((SERVER_HOST, SERVER_PORT))
print(f'Connected to server at {SERVER_HOST}:{SERVER_PORT}')

# If reverse mode is enabled, send 'R' to the server; otherwise, send 'N'
if args.reverse:
    client_socket.sendall(b'R')
    print("Reverse mode: server will send data to client.")
    
    start_time = time.time()
    total_data_received = 0
    log_time = start_time
    
    try:
        # Receive data for the specified duration
        while time.time() - start_time < DOWNLOAD_DURATION:
            data = client_socket.recv(BUFFER_SIZE)
            if not data:
                break
            total_data_received += len(data)
            
            # Log transfer progress
            if time.time() - log_time >= LOG_INTERVAL:
                elapsed_time = time.time() - start_time
                throughput_mbps = (total_data_received * 8 / (1024 * 1024)) / elapsed_time
                print(f"[Client] Time: {elapsed_time:.2f}s, Data received: {total_data_received / (1024 * 1024):.2f} MB, Throughput: {throughput_mbps:.2f} Mbps")
                log_time = time.time()

    except KeyboardInterrupt:
        print("Download interrupted.")
    finally:
        client_socket.close()
    
    # Report the total data received
    print(f'Total data received: {total_data_received / (1024 * 1024):.2f} MB')

else:
    client_socket.sendall(b'N')
    print("Normal mode: client will send data to server.")
    
    start_time = time.time()
    total_data_sent = 0
    log_time = start_time
    
    try:
        # Send data for the specified duration
        while time.time() - start_time < DOWNLOAD_DURATION:
            client_socket.sendall(DATA)
            total_data_sent += len(DATA)
            
            # Log transfer progress
            if time.time() - log_time >= LOG_INTERVAL:
                elapsed_time = time.time() - start_time
                throughput_mbps = (total_data_sent * 8 / (1024 * 1024)) / elapsed_time
                print(f"[Client] Time: {elapsed_time:.2f}s, Data sent: {total_data_sent / (1024 * 1024):.2f} MB, Throughput: {throughput_mbps:.2f} Mbps")
                log_time = time.time()

    except KeyboardInterrupt:
        print("Data transfer interrupted.")
    finally:
        client_socket.close()
    
    # Report the total data sent
    print(f"Total data sent: {total_data_sent / (1024 * 1024):.2f} MB")

File: client_side.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <time.h>

#define BUFFER_SIZE 128 * 1024  // Default buffer size (128 KB)
#define SERVER_PORT 5201

void handle_error(const char *msg) {
    perror(msg);
    exit(EXIT_FAILURE);
}

void reverse_mode(int client_socket, size_t buffer_size) {
    char buffer[buffer_size];
    ssize_t bytes_received;
    size_t total_data_received = 0;
    clock_t start_time = clock();

    printf("Reverse Mode: Receiving data from the server.\n");

    while ((bytes_received = recv(client_socket, buffer, buffer_size, 0)) > 0) {
        total_data_received += bytes_received;
    }

    if (bytes_received < 0) {
        handle_error("recv failed");
    }

    double elapsed_time = (double)(clock() - start_time) / CLOCKS_PER_SEC;
    double throughput_mbps = (total_data_received * 8) / (1024.0 * 1024.0 * elapsed_time);

    printf("Total data received: %.2f MB, Throughput: %.2f Mbps\n",
           total_data_received / (1024.0 * 1024.0), throughput_mbps);
}

void normal_mode(int client_socket, int iterations, int sleep_duration, int constant_rate, int phase_time,
                 int target_rate, int rate_based_phase, int time_based_phase, int bytes, int time_sec,
                 size_t buffer_size) {
    char data[buffer_size];
    memset(data, 'X', buffer_size);

    for (int i = 0; i < iterations; i++) {
        size_t total_data_sent = 0;
        clock_t start_time = clock();
        double max_throughput_mbps = 0;

        printf("[Client] Iteration %d started.\n", i + 1);

        // Phase 1: Increasing Phase
        if (constant_rate) {
            if (rate_based_phase && target_rate) {
                printf("[Client] Increasing phase based on target rate.\n");

                while (max_throughput_mbps < target_rate) {
                    if (send(client_socket, data, buffer_size, 0) < 0) {
                        handle_error("send failed");
                    }
                    total_data_sent += buffer_size;
                    double elapsed_time = (double)(clock() - start_time) / CLOCKS_PER_SEC;
                    max_throughput_mbps = (total_data_sent * 8) / (1024.0 * 1024.0 * elapsed_time);
                }
            } else if (time_based_phase) {
                printf("[Client] Increasing phase based on time.\n");

                clock_t phase_end_time = start_time + time_based_phase * CLOCKS_PER_SEC;
                while (clock() < phase_end_time) {
                    if (send(client_socket, data, buffer_size, 0) < 0) {
                        handle_error("send failed");
                    }
                    total_data_sent += buffer_size;
                }
            }

            // Phase 2: Constant Rate Phase
            clock_t constant_phase_end = clock() + phase_time * CLOCKS_PER_SEC;
            size_t bytes_per_second = max_throughput_mbps * 1024 * 1024 / 8;
            size_t bytes_per_interval = bytes_per_second * 0.01;

            printf("[Client] Constant rate phase started.\n");
            while (clock() < constant_phase_end) {
                size_t chunk_size = buffer_size < bytes_per_interval ? buffer_size : bytes_per_interval;
                if (send(client_socket, data, chunk_size, 0) < 0) {
                    handle_error("send failed");
                }
                total_data_sent += chunk_size;

                usleep(10000);  // Sleep for 10 ms
            }
        } else {
            // Non-constant rate, either bytes or time-based transfer
            if (bytes > 0) {
                printf("[Client] Sending %d bytes.\n", bytes);
                size_t bytes_to_send = bytes;
                while (bytes_to_send > 0) {
                    size_t chunk_size = buffer_size < bytes_to_send ? buffer_size : bytes_to_send;
                    if (send(client_socket, data, chunk_size, 0) < 0) {
                        handle_error("send failed");
                    }
                    total_data_sent += chunk_size;
                    bytes_to_send -= chunk_size;
                }
            } else if (time_sec > 0) {
                printf("[Client] Sending data for %d seconds.\n", time_sec);
                clock_t phase_end_time = clock() + time_sec * CLOCKS_PER_SEC;
                while (clock() < phase_end_time) {
                    if (send(client_socket, data, buffer_size, 0) < 0) {
                        handle_error("send failed");
                    }
                    total_data_sent += buffer_size;
                }
            }
        }

        // Log transfer progress
        double elapsed_time = (double)(clock() - start_time) / CLOCKS_PER_SEC;
        double throughput_mbps = (total_data_sent * 8) / (1024.0 * 1024.0 * elapsed_time);

        printf("[Client] Iteration %d completed, Data sent: %.2f MB, Throughput: %.2f Mbps\n",
               i + 1, total_data_sent / (1024.0 * 1024.0), throughput_mbps);

        // Sleep between iterations
        if (sleep_duration > 0) {
            printf("[Client] Sleeping for %d seconds.\n", sleep_duration);
            sleep(sleep_duration);
        }
    }
}

int main(int argc, char *argv[]) {
    if (argc < 2) {
        fprintf(stderr, "Usage: %s -s <server_ip> [options]\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    char *server_ip = NULL;
    int server_port = SERVER_PORT;
    int reverse = 0, iterations = 1, sleep_duration = 0, constant_rate = 0;
    int phase_time = 5, target_rate = 0, rate_based_phase = 0, time_based_phase = 0, bytes = 0, time_sec = 0;
    size_t buffer_size = BUFFER_SIZE;

    // Parse command-line arguments
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "-s") == 0 && i + 1 < argc) {
            server_ip = argv[++i];
        } else if (strcmp(argv[i], "-p") == 0 && i + 1 < argc) {
            server_port = atoi(argv[++i]);
        } else if (strcmp(argv[i], "-R") == 0) {
            reverse = 1;
        } else if (strcmp(argv[i], "--iterations") == 0 && i + 1 < argc) {
            iterations = atoi(argv[++i]);
        } else if (strcmp(argv[i], "--sleep") == 0 && i + 1 < argc) {
            sleep_duration = atoi(argv[++i]);
        } else if (strcmp(argv[i], "--constant_rate") == 0) {
            constant_rate = 1;
        } else if (strcmp(argv[i], "--phase_time") == 0 && i + 1 < argc) {
            phase_time = atoi(argv[++i]);
        } else if (strcmp(argv[i], "--target_rate") == 0 && i + 1 < argc) {
            target_rate = atoi(argv[++i]);
        } else if (strcmp(argv[i], "--rate_based_phase") == 0) {
            rate_based_phase = 1;
        } else if (strcmp(argv[i], "--time_based_phase") == 0 && i + 1 < argc) {
            time_based_phase = atoi(argv[++i]);
        } else if (strcmp(argv[i], "--bytes") == 0 && i + 1 < argc) {
            bytes = atoi(argv[++i]);
        } else if (strcmp(argv[i], "--time") == 0 && i + 1 < argc) {
            time_sec = atoi(argv[++i]);
        } else if (strcmp(argv[i], "--buffer_size") == 0 && i + 1 < argc) {
            buffer_size = atoi(argv[++i]);
        }
    }

    if (!server_ip) {
        fprintf(stderr, "Server IP address is required.\n");
        exit(EXIT_FAILURE);
    }

    int client_socket;
    struct sockaddr_in server_addr;

    // Create a TCP socket
    if ((client_socket = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
        handle_error("socket creation failed");
    }

    // Set server address
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(server_port);
    if (inet_pton(AF_INET, server_ip, &server_addr.sin_addr) <= 0) {
        handle_error("invalid address");
    }

    // Connect to the server
    if (connect(client_socket, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
        handle_error("connection failed");
    }

    printf("Connected to server at %s:%d\n", server_ip, server_port);

    // Send mode to the server ('N' for Normal, 'R' for Reverse)
    char mode = reverse ? 'R' : 'N';
    if (send(client_socket, &mode, 1, 0) < 0) {
        handle_error("send failed");
    }

    if (reverse) {
        reverse_mode(client_socket, buffer_size);
    } else {
        normal_mode(client_socket, iterations, sleep_duration, constant_rate, phase_time,
                    target_rate, rate_based_phase, time_based_phase, bytes, time_sec, buffer_size);
    }

    // Close the client socket
    close(client_socket);
    printf("Connection closed.\n");

    return 0;
}
